File: event_counter.ts
interface GraphType {
    [key: string]: number
}


class EventCounter {
    private graph: GraphType = {};

    public add = (event: string): void => {
        const count = this.graph[event] || 0;
        this.graph[event] = count + 1;
    }

    public topK = (k: number): Array<string> => {
        const eventList: Array<{event: string, count: number}> = [];
        Object.entries(this.graph).forEach(([event, count]) => {
            eventList.push({
                event,
                count,
            });
        })
        eventList.sort((itemA, itemB) => itemB.count - itemA.count);
        const results = eventList.map((item) => item.event);
        return results.slice(0, k);
    } 
}

const ec = new EventCounter()

// count: checkout 3, login 2, purchase 1
ec.add("purchase")
ec.add("login")
ec.add("checkout")
ec.add("login")
ec.add("checkout")
ec.add("checkout")

// expected: checkout, login
console.log(ec.topK(2))

// # expected: checkout, login, purchase
console.log(ec.topK(3))

File: event_counter.py
from collections import defaultdict, Counter
import heapq

class EventCounter:
    def __init__(self) -> None:
        self.counts = defaultdict(int)

    def add(self, event: str) -> None:
        self.counts[event] += 1

    def top_k(self, k: int) -> list[str]:
        heap = []
        for event, freq in self.counts.items():
            # heapq is a min heap... smallest will be first to pop
            heapq.heappush(heap, (-freq, event))

        result = []
        for _ in range(k):
            _, event = heapq.heappop(heap)
            result.append(event)
        return result


# class EventCounter:
#     def __init__(self) -> None:
#         self.counts = Counter([])

#     def add(self, event: str) -> None:
#         self.counts.update([event])

#     def top_k(self, k: int) -> list[str]:
#         return self.counts.most_common(k)



ec = EventCounter()

# count: checkout 3, login 2, purchase 1
ec.add("purchase")
ec.add("login")
ec.add("checkout")
ec.add("login")
ec.add("checkout")
ec.add("checkout")

# expected: checkout, login
print(ec.top_k(2))
# expected: checkout, login, purchase
print(ec.top_k(3))
